internal/controlplane/common.go

```diff
@@ -67,13 +67,45 @@ func getProviderFromRequestOrDefault(
 			len(providers))
 	}
 
+	return findProvider(in.GetContext().GetProvider(), providers)
+}
+
+func listProvidersOrInferDefault(
+	ctx context.Context,
+	store db.Store,
+	in HasProtoContext,
+	projectId uuid.UUID,
+) ([]db.Provider, error) {
+	// Allows us to take into account the hierarchy to find the provider
+	parents, err := store.GetParentProjects(ctx, projectId)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "cannot retrieve parent projects: %s", err)
+	}
+	providers, err := store.ListProvidersByProjectID(ctx, parents)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "cannot retrieve providers: %s", err)
+	}
+	// if we do not have a provider name, check if we can infer it
+	if in.GetContext().GetProvider() == "" {
+		return providers, nil
+	}
+
+	prov, err := findProvider(in.GetContext().GetProvider(), providers)
```

Comment:
Again, not for this PR, but this feels massively inefficient from a data transfer point of view -- we ignore the provider argument and fetch _all possible providers_, only to filter through all of them looking for one that matches the provider argument, rather than passing the provider as a `sql.NullString` to a `store` method.

But, don't change that for this PR, we should do it in a separate refactor.